alu_u32/src/mul/stark.rs

```diff
@@ -21,23 +19,26 @@ where
         let next: &Mul32Cols<AB::Var> = main.row_slice(1).borrow();
 
         // Limb weights modulo 2^32
-        let base_m = [1, 1 << 8, 1 << 16, 1 << 24].map(AB::Expr::from_canonical_u32);
+        let base_m32 = [1, 1 << 8, 1 << 16, 1 << 24].map(AB::Expr::from_canonical_u32);
+
+        // Limb weights modulo 2^16
+        let base_m16 = [1, 1 << 8].map(AB::Expr::from_canonical_u32);
 
         // Partially reduced summation of input product limbs (mod 2^32)
-        let pi = pi_m::<4, AB>(&base_m, local.input_1, local.input_2);
+        let pi = pi_m::<4, AB::Var, AB::Expr>(&base_m32, local.input_1, local.input_2);
 
         // Partially reduced summation of output limbs (mod 2^32)
-        let sigma = sigma_m::<4, AB>(&base_m, local.output);
+        let sigma = sigma_m::<4, AB::Var, AB::Expr>(&base_m32, local.output);
 
         // Partially reduced summation of input product limbs (mod 2^16)
-        let pi_prime = pi_m::<2, AB>(&base_m[..2], local.input_1, local.input_2);
+        let pi_prime = pi_m::<2, AB::Var, AB::Expr>(&base_m16, local.input_1, local.input_2);
 
         // Partially reduced summation of output limbs (mod 2^16)
-        let sigma_prime = sigma_m::<2, AB>(&base_m[..2], local.output);
+        let sigma_prime = sigma_m::<2, AB::Var, AB::Expr>(&base_m16, local.output);
 
         // Congruence checks
-        builder.assert_eq(pi - sigma, local.r * AB::Expr::TWO);
-        builder.assert_eq(pi_prime - sigma_prime, local.s * base_m[2].clone());
+        builder.assert_eq(pi - sigma, local.r * AB::Expr::from_wrapped_u64(1 << 32));
+        builder.assert_eq(pi_prime - sigma_prime, local.s * base_m32[2].clone());
 
         // Range check counter
         builder
```

Comment:
Look like this enforces a range of `1 ..= 2^10`, should it be `0 .. 2^10`?

Can also use `assert_bool` on the counter diff